src/python/dxpy/bindings/apollo/ValidateJSONbySchema.py

```diff
@@ -0,0 +1,165 @@
+                    "nested_key2": {"type": str}
+                }
+            }
+        },
+        "id": {
+            "type": list,
+        },
+        "conflicting_keys": [["key1", "key2"]],
+        "dependent_conditional_keys": {
+            "list_key_example": ["key1", "key2"]
+        }
+    }
+
+    Key types can be defined via the "type" key
+
+    Required keys can be defined via the "required" key
+
+    Conflicting keys can be defined via the "conflicting_keys" key, where the value is a list of lists of keys that are mutually exclusive.
+    Conflicting keys can be defined at the level of the schema for the entire JSON or at key-level.
+    In the example above, within "key2" the keys "xyz" and "abc" are mutually exclusive and cannot be present together.
+    Similarly, at the level of the whole JSON, "key1" and "key2" are mutually exclusive and cannot be present together.
+
+    Conditional key combinations can be defined via the dependent_conditional_keys at the level of the schema. These should be defined as a dict where the value is a list.
+    In the example above, dependent_conditional_keys is defined as {"list_key_example": ["key1", "key2"]}, which means whenever list_key_example is present, either "key1"
+    or "key2" has to be also present.
+
+    Currently the maximum level of nestedness supported for validation is as deep as shown in the example above via the "list_key_example" key.
+
+    This is an example of a valid JSON that satisfies all the conditions defined in the schema above:
+
+    {
+        "key1": {
+            "name": "test",
+        },
+        "list_key_example": [
+            {"nested_key1": "1", "nested_key2": "2"},
+            {"nested_key1": "3", "nested_key2": "4"},
+            {"nested_key1": "5"}
+        ],
+        "id": ["id_1", "id_2"]
+    }
+
+    """
+
+    def __init__(self, schema):
+        self.schema = schema
+
+    def validate(self, input_json):
+        for key, value in self.schema.items():
+            if key in input_json:
+                self.validate_properties(value.get("properties", {}), input_json[key])
+
+                if value.get("type") == list:
+                    self.validate_list_items(
+                        value.get("items", {}), input_json[key], key
+                    )
+
+                # Check for incompatible/conflicting subkeys defined at the key-level
+                if value.get("conflicting_keys"):
+                    self.check_incompatible_subkeys(input_json[key], key)
+
+        self.check_incompatible_keys(input_json)
+        self.check_dependent_key_combinations(input_json)
+
+    def validate_properties(self, properties, input_dict):
+        for key, value in properties.items():
+            if key not in input_dict and value.get("required"):
+                err_exit(
+                    "Required key '{}' was not found in the input JSON.".format(key)
+                )
```

Comment:
For this and following `err_exit` calls I would recommend to utilize the delegator pattern. In the class constructor you would accept a method object which will be called instead of `err_exit`. Then the actual `err_exit` call would be defined outside, hopefully closer to the main flow. Also easily replaceable for the whole class as a behaviour.

For example the usage can get as easy as:
```
validator = JSONValidator(schema, err_exit)
``` 
But would allow for easy modification should the error reporting change:
```
def error_handler(message):
    LOG.error(message)
    err_exit(message)
validator = JSONValidator(schema, error_handler)
```

Reply:
That's a very nice idea. I'll implement to my classes as well

Reply:
Done. Please see https://github.com/dnanexus/dx-toolkit/pull/1157/commits/9b336e65ec0c508aaf1619ac09d98e3208fb42a3